src/compiler/checker.ts

```diff
@@ -17789,7 +17789,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function maybeCloneTypeParameter(p: TypeParameter) {
         const constraint = getConstraintOfTypeParameter(p);
-        return constraint && (isGenericObjectType(constraint) || isGenericIndexType(constraint)) ? cloneTypeParameter(p) : p;
+        return constraint && couldContainTypeVariables(constraint) ? cloneTypeParameter(p) : p;

```

Comment:
I keep thinking that perhaps there is a better check for this kind of thing - but I honestly don't know what it is. I already had a need for this a couple of times and I couldn't find anything else in the codebase that would already test this kind of stuff in a different way.

Reply:
This function was introduced back in #31455; I suspect that it was a targeted fix and personally I'm partial to what you're changing it to (it makes sense to me as an optimization; if we always called clone, all tests pass, it's just slower).

@weswigham